test/state/rlp.cpp

```diff
@@ -0,0 +1,74 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "rlp.hpp"
+#include "state.hpp"
+
+namespace evmone::state
+{
+// FIXME: Same function already defined in statetest_logs_hash.cpp
+/// Defines how to RLP-encode a Log. This is only needed to encode transaction receipt.
+inline bytes rlp_encode(const Log& log)
+{
+    return rlp::encode_tuple(log.addr, log.topics, log.data);
+}
+}  // namespace evmone::state
+
+namespace evmone::rlp
+{
+
+namespace
+{
+inline bytes legacy_tx_to_rlp(uint64_t nonce, const intx::uint256& gas_price, uint64_t gas_limit,
+    const bytes_view& to, const intx::uint256& value, const bytes_view& data, uint8_t v,
+    const intx::uint256& r, const intx::uint256& s)
+{
+    assert(to.size() == 20);
+    // rlp [nonce, gas_price, gas_limit, to, value, data, v, r, s];
+    return rlp::encode_tuple(nonce, gas_price, gas_limit, to, value, data, v, r, s);
+}
+
+inline bytes eip1559_tx_to_rlp(uint64_t chain_id, uint64_t nonce,
+    const intx::uint256& max_priority_gas_price, const intx::uint256& max_gas_price,
+    uint64_t gas_limit, const bytes_view& to, const intx::uint256& value, const bytes_view& data,
+    const state::AccessList& /*access_list*/, uint8_t v, const intx::uint256& r,
+    const intx::uint256& s)
+{
+    // TODO: Implement AccessList encoding
+    assert(to.size() == 20);
+    assert(v == 1 || v == 0);
+    // tx_type +
+    // rlp [chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit,
+    // to, value, data, access_list, sig_parity, r, s];
+    return bytes({0x02}) +  // Transaction type (eip1559 type == 2)
+           rlp::encode_tuple(chain_id, nonce, max_priority_gas_price, max_gas_price, gas_limit, to,
+               value, data, std::vector<uint8_t>(), v, r, s);
+}
+
+}  // namespace
+
+bytes encode(const state::Transaction& tx)
+{
+    using namespace evmone;
+
+    if (tx.kind == state::Transaction::Kind::legacy)
+        return legacy_tx_to_rlp(tx.nonce, tx.max_gas_price, static_cast<uint64_t>(tx.gas_limit),
+            tx.to.has_value() ? bytes_view(tx.to.value()) : bytes(), tx.value, tx.data, tx.v, tx.r,
+            tx.s);
+    else
+        return eip1559_tx_to_rlp(tx.chain_id, tx.nonce, tx.max_priority_gas_price, tx.max_gas_price,
+            static_cast<uint64_t>(tx.gas_limit),
+            tx.to.has_value() ? bytes_view(tx.to.value()) : bytes(), tx.value, tx.data,
+            tx.access_list, tx.v, tx.r, tx.s);
+}
+
+[[nodiscard]] bytes encode(const state::TransactionReceipt& tx_receipt)
```

Comment:
Why not call these `rlp_encode` to stay with the overload?

Reply:
I moved all `rlp_encode` functions decl to `state.hpp` and remove the mess I made in rlp.hpp. But do not want to put the implementations to header. So they went to `state.cpp` problem is that now they cannot be inlined.